src/catalog/src/remote/cached_client.rs

```diff
@@ -0,0 +1,194 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::any::Any;
+use std::sync::Arc;
+use std::time::Duration;
+
+use common_telemetry::timer;
+use meta_client::client::MetaClient;
+use moka::future::{Cache, CacheBuilder};
+
+use super::MetaKvBackend;
+use crate::error::Error::NoSuchKey;
+use crate::error::{Error, GenericSnafu, NoSuchKeySnafu, Result};
+use crate::metrics::{METRIC_CATALOG_KV_GET, METRIC_CATALOG_KV_REMOTE_GET};
+use crate::remote::{CachedKvBackend, Kv, KvBackend, KvBackendRef, ValueIter};
+
+pub struct CachedCatalogKvBackend {
+    inner_kv_backend: KvBackendRef,
+    cache: Arc<Cache<Vec<u8>, Kv>>,
+}
+
+#[async_trait::async_trait]
+impl KvBackend for CachedCatalogKvBackend {
+    fn as_any(&self) -> &(dyn Any + Send + Sync) {
+        self
+    }
+
+    fn range<'a, 'b>(&'a self, key: &[u8]) -> ValueIter<'b, Error>
+    where
+        'a: 'b,
+    {
+        self.inner_kv_backend.range(key)
+    }
+
+    async fn get(&self, key: &[u8]) -> Result<Option<Kv>> {
+        let _timer = timer!(METRIC_CATALOG_KV_GET);
+
+        let init = async {
+            let _timer = timer!(METRIC_CATALOG_KV_REMOTE_GET);
+
+            let kv = self.inner_kv_backend.get(key).await;
+            convert(kv)
+        };
+
+        let schema_provider = self.cache.try_get_with_by_ref(key, init).await;
+
+        re_convert(schema_provider)
+    }
+
+    async fn set(&self, key: &[u8], val: &[u8]) -> Result<()> {
+        self.inner_kv_backend.set(key, val).await
+    }
+
+    async fn delete(&self, key: &[u8]) -> Result<()> {
+        self.delete_range(key, &[]).await
+    }
+
+    async fn delete_range(&self, key: &[u8], end: &[u8]) -> Result<()> {
+        self.inner_kv_backend.delete_range(key, end).await
+    }
+
+    async fn compare_and_set(
+        &self,
+        key: &[u8],
+        expect: &[u8],
+        val: &[u8],
+    ) -> Result<std::result::Result<(), Option<Vec<u8>>>> {
+        self.inner_kv_backend
+            .compare_and_set(key, expect, val)
+            .await
+    }
```

Comment:
The `delete_range` and `set` should invalidate the cache too.

Reply:
Cache expiration in the current design refers to the invalidation of business-level metadata, and this is a distributed behavior. Although it is not wrong for the current node to handle expired caches on its own, it can also be omitted.

Reply:
The correct way to invalidate cache is to rely on distributed procedures that require distributed scheduling in metasrv to invalidate the cache.
For example:
`alter table` in procedure -> metasrv broadcast invalidate cache instruction via mail box -> All frontend.invaliddate_table_cache(table_name) -> cache.invalidate(table_keys[])